test/Tester/GrainCallFilterTests.cs

```diff
@@ -0,0 +1,199 @@
+using System;
+using System.Globalization;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Orleans.Configuration;
+using Orleans.Providers;
+using Orleans.Runtime;
+using Orleans.Runtime.Configuration;
+using Orleans.TestingHost;
+using TestExtensions;
+using UnitTests.GrainInterfaces;
+using Xunit;
+
+namespace UnitTests.General
+{
+    [TestCategory("BVT"), TestCategory("GrainCallFilter")]
+    public class GrainCallFilterTests : TestClusterPerTest
+    {
+        public override TestCluster CreateTestCluster()
+        {
+            var options = new TestClusterOptions(2);
+            options.ClusterConfiguration.AddMemoryStorageProvider("Default");
+            options.ClusterConfiguration.AddMemoryStorageProvider("PubSubStore");
+            options.ClusterConfiguration.AddSimpleMessageStreamProvider("SMSProvider");
+            options.ClientConfiguration.AddSimpleMessageStreamProvider("SMSProvider");
+            options.ClusterConfiguration.Globals.RegisterBootstrapProvider<PreInvokeCallbackBootrstrapProvider>(
+                "PreInvokeCallbackBootrstrapProvider");
+            options.ClusterConfiguration.UseStartupType<SiloInvokerTestStartup>();
+            return new TestCluster(options);
+        }
+
+        /// <summary>
+        /// Ensures that grain call filters are invoked around method calls in the correct order.
+        /// </summary>
+        /// <returns>A <see cref="Task"/> representing the work performed.</returns>
+        [Fact]
+        public async Task GrainCallFilter_Order_Test()
+        {
+            var grain = this.GrainFactory.GetGrain<IGrainCallFilterTestGrain>(random.Next());
+
+            // This grain method reads the context and returns it
+            var context = await grain.GetRequestContext();
+            Assert.NotNull(context);
+            Assert.Equal("grain!", context);
```

Comment:
lol, this is kind of clever, although it's probably much more readable if this would end up concatenating "123456"....